libwnbd/nbd_daemon.cpp

```diff
@@ -0,0 +1,569 @@
+        LogError("Could not allocate %d bytes.", PreallocatedWBuffSz);
+        return ERROR_NOT_ENOUGH_MEMORY;
+    }
+    PreallocatedRBuffSz = WNBD_DEFAULT_MAX_TRANSFER_LENGTH;
+    PreallocatedRBuff = (PVOID) calloc(1, PreallocatedRBuffSz);
+    if (!PreallocatedRBuff) {
+        LogError("Could not allocate %d bytes.", PreallocatedRBuffSz);
+        return ERROR_NOT_ENOUGH_MEMORY;
+    }
+
+    DWORD Err = ConnectNbdServer(
+        WnbdProps.NbdProperties.Hostname,
+        WnbdProps.NbdProperties.PortNumber);
+    if (Err) {
+        return Err;
+    }
+    Err = SetTcpFlags(Socket);
+    if (Err) {
+        return Err;
+    }
+
+    UINT16 NbdFlags = 0;
+    if (!WnbdProps.NbdProperties.Flags.SkipNegotiation) {
+        uint64_t DiskSize = 0;
+
+        Err = NbdNegotiate(Socket, &DiskSize, &NbdFlags,
+                           WnbdProps.NbdProperties.ExportName,
+                           NBD_FLAG_FIXED_NEWSTYLE);
+        if (Err) {
+            LogError("NBD negotiation failed.");
+            return Err;
+        }
+
+        WnbdProps.BlockCount = DiskSize / WnbdProps.BlockSize;
+    }
+
+    WnbdProps.Flags.ReadOnly |= CHECK_NBD_READONLY(NbdFlags);
+    WnbdProps.Flags.UnmapSupported |= CHECK_NBD_SEND_TRIM(NbdFlags);
+    WnbdProps.Flags.FlushSupported |= CHECK_NBD_SEND_FLUSH(NbdFlags);
+    WnbdProps.Flags.FUASupported |= CHECK_NBD_SEND_FUA(NbdFlags);
+
+    if (!WnbdProps.BlockCount ||
+            WnbdProps.BlockCount > ULLONG_MAX / WnbdProps.BlockSize) {
+        LogError("Invalid block size or block count. "
+                 "Block size: %d. Block count: %lld.",
+                 WnbdProps.BlockSize, WnbdProps.BlockCount);
+        return ERROR_INVALID_PARAMETER;
+    }
+
+    LogInfo("Retrieved NBD flags: %d. Read-only: %d, TRIM enabled: %d, "
+            "FLUSH enabled: %d, FUA enabled: %d.",
+            NbdFlags,
+            WnbdProps.Flags.ReadOnly,
+            WnbdProps.Flags.UnmapSupported,
+            WnbdProps.Flags.FlushSupported,
+            WnbdProps.Flags.FUASupported);
+
+    ReplyDispatcher = std::thread(&NbdDaemon::NbdReplyWorker, this);
+
+    Err = WnbdCreate(
+        &WnbdProps, (const PWNBD_INTERFACE) &WnbdInterface,
+        this, &WnbdDisk);
+    if (Err) {
+        return Err;
+    }
+
+    // We currently use a single NBD connection, which is why
+    // we'll stick with a single WNBD worker thread.
+    Err = WnbdStartDispatcher(WnbdDisk, 1);
+    if (Err) {
+        return Err;
+    }
+
+    LogInfo("NBD mapping initialized successfully.");
+    return 0;
+}
+
+DWORD NbdDaemon::Start()
+{
+    DWORD Retval = TryStart();
+    if (Retval) {
+        LogError("NBD daemon failed to initialize, attempting cleanup.");
+        Shutdown(true);
+        Wait();
+    }
+    return Retval;
+}
+
+DWORD NbdDaemon::Shutdown(bool HardRemove)
+{
+    std::unique_lock<std::mutex> Lock(ShutdownLock);
+
+    if (!Terminated) {
+        TerminateInProgress = true;
+        DWORD Err = DisconnectNbd();
```

Comment:
TODO: we should remove the NBD connection after the WNBD one, allowing pending IO to finish.